README.md
# ABOUT

<!-- ## Homepage -->


## Description
This is a simple `JavaScript` HTML and CSS preprocessor.  
This was originally a fork of `https://github.com/krasimir/absurd.git` before I personalized it.


<!-- ## Features -->


<!-- ## Core Dependancies -->


<!-- ## Inspirations -->


<!-- ## Screen Shots -->


# INSTALLATION


<!-- ## Prerequisites -->


## Install
I do not intend to publish this package to the `npm` package registry. If you have `verdaccio` installed in your system simply:

```bash
    git clone -b forked_dev https://github.com/kilisio/preprocess.git
    cd preprocess
    npm install
    npm run dist
    npm run pub_lib
```

The above will publish the package to `verdaccio` if you have it configured as your default package publishing registry.  
Then simply install and add it to your project script:

```bash
    npm install @kilisio/preprocess
```

```javascript
    import preprocess from '@kilisio/preprocess';
```

The package scope `@kilisio` and other package details may be changed and configured in the `lib.config.js` script before building the library.


## Usage

```javascript
var test_html = {
    div: 'hello world'
};

var test_css = {
    div:{
        margin: 0,
        padding: 0,
        position: 'absolute',
        width: '224px',
        height: '30px',
		'background-color': '#1a0a00',
		'padding-left': '5px',
        '.ul_nav': {
			margin: 0,
			padding: 0,
			'list-style': 'none'
		},
        '.ul_nav li': {
			margin: 0,
			padding: 0,
			display: "inline"
		},
        '.link_group': {
            margin: 0,
            padding: 0,
            display: 'inline-block',
			':hover':{
                '.link_nav': {
                    color: '#ff6600'
                },
                '.nav_img': {
                    filter: 'invert(92%) sepia(13%) saturate(682%) hue-rotate(311deg) brightness(107%) contrast(101%)'
                },
                '.nav_img_bg': {
                    'background-color': '#1a0a00'
                }
			}
        },
        '.nav_img_bg': {
			margin: 0,
			padding: 0,
			width: '30px',
			height: '30px',
			'margin-right': '3px',
			'background-color': '#fff0e6',
			display: 'inline-block',
			'border-radius': '50%',
			'vertical-align': 'middle'
		},
        '.nav_img': {
			margin: '6px',
			padding: 0,
			width: '18px',
			height: '18px',
            filter: 'invert(6%) sepia(76%) saturate(641%) hue-rotate(345deg) brightness(95%) contrast(104%)'
		},
        '.link_nav': {
			margin: 0,
			padding: 0,
			'text-decoration': 'none',
			color: '#fff0e6',
			'font-family': 'forced_squaremedium',
			'font-size': '28px',
			'margin-right': '3px',
			'vertical-align': 'middle',
            outline: 'none',
            ':active :focus': {
                outline: 'none'
            }
		}
    }
};


// var test_component = {
//    html: test_html,
//    css: test_css
// };

preprocess(function(api){
    api.add(test_css).compile(function(err, result){
        console.log(result);
    });
    api.morph('html').add(test_html).compile(function(err, result){
        console.log(result);
    });
    // api.add(test_component.css).compile(function(err, result){
    //     console.log(result);
    // });
    // api.morph('html').add(test_component.html).compile(function(err, result){
    //     console.log(result);
    // });
});

```
output:  
```
div, div .ul_nav, div .ul_nav li, div .link_group, div .nav_img_bg, div .link_nav {
  margin: 0;
}
div, div .ul_nav, div .ul_nav li, div .link_group, div .nav_img_bg, div .nav_img, div .link_nav {
  padding: 0;
}
div {
  position: absolute;
  width: 224px;
  padding-left: 5px;
}
div, div .nav_img_bg {
  height: 30px;
}
div, div .link_group:hover .nav_img_bg {
  background-color: #1a0a00;
}
div .ul_nav {
  list-style: none;
}
div .ul_nav li {
  display: inline;
}
div .link_group, div .nav_img_bg {
  display: inline-block;
}
div .link_group:hover .link_nav {
  color: #ff6600;
}
div .link_group:hover .nav_img {
  filter: invert(92%) sepia(13%) saturate(682%) hue-rotate(311deg) brightness(107%) contrast(101%);
}
div .nav_img_bg {
  width: 30px;
  background-color: #fff0e6;
  border-radius: 50%;
}
div .nav_img_bg, div .link_nav {
  margin-right: 3px;
  vertical-align: middle;
}
div .nav_img {
  margin: 6px;
  width: 18px;
  height: 18px;
  filter: invert(6%) sepia(76%) saturate(641%) hue-rotate(345deg) brightness(95%) contrast(104%);
}
div .link_nav {
  text-decoration: none;
  color: #fff0e6;
  font-family: forced_squaremedium;
  font-size: 28px;
}
div .link_nav, div .link_nav:active :focus {
  outline: none;
}

<div>hello world</div>
```

<!-- # DOCUMENTATION -->


<!-- # CONTRIBUTION -->


<!-- ## Features, Issues and Fixes -->


<!-- ## Author -->


<!-- ## Contributors -->


<!-- ## Backers -->


<!-- ## Sponsors -->


<!-- # ATTRIBUTION -->


<!-- # SUPPORT -->


# LICENSE
Copyright © 2019 Patrick Kilisio.
This project is MIT licensed.



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.3.10"></a>
## [0.3.10](https://github.com/kilisio/preprocess/compare/v0.3.9...v0.3.10) (2019-12-27)


### Bug Fixes

* **api_morph.js:** removed component preprocessing option ([081bb0d](https://github.com/kilisio/preprocess/commit/081bb0d))



<a name="0.3.9"></a>
## 0.3.9 (2019-12-27)


### Bug Fixes

* ***.js:** changed scripts to es import/export syntax ([2946321](https://github.com/kilisio/preprocess/commit/2946321))
* **api_add.js:** fixed semantic errors ([2e1f4c2](https://github.com/kilisio/preprocess/commit/2e1f4c2))
* **api_compile.js:** fixed semantic errors ([970a9eb](https://github.com/kilisio/preprocess/commit/970a9eb))
* **api_morph.js:** fixed component and html api hook expression ([9a4c3b4](https://github.com/kilisio/preprocess/commit/9a4c3b4))
* **api_morph.js:** fixed semantic errors ([f2eab3b](https://github.com/kilisio/preprocess/commit/f2eab3b))
* **api.js:** added additional api methods ([bfe1a0c](https://github.com/kilisio/preprocess/commit/bfe1a0c))
* **api.js:** typo ([33cebac](https://github.com/kilisio/preprocess/commit/33cebac))
* **func_analyzer.js:** fixed semantic syntax errors ([21d2839](https://github.com/kilisio/preprocess/commit/21d2839))
* **func_component.js:** fixed semantic syntax errors ([e5f0617](https://github.com/kilisio/preprocess/commit/e5f0617))
* **func_css.js:** fixed semantic syntax errors ([b8e7a6c](https://github.com/kilisio/preprocess/commit/b8e7a6c))
* **func_html.js:** corrected semantic errors ([9f3f95b](https://github.com/kilisio/preprocess/commit/9f3f95b))
* **func_morph.js:** fixed for loop syntax ([d2b05bc](https://github.com/kilisio/preprocess/commit/d2b05bc))
* **func_prefixes.js:** fixed semantic errors ([dfefab1](https://github.com/kilisio/preprocess/commit/dfefab1))
* **func_template.js:** fixed some semantic errors ([88a0f1e](https://github.com/kilisio/preprocess/commit/88a0f1e))
* **func_transform.js:** fixed for loop syntax; replaced var with let ([38062bb](https://github.com/kilisio/preprocess/commit/38062bb))
* package.json & .snyk to reduce vulnerabilities ([4d91cb4](https://github.com/kilisio/preprocess/commit/4d91cb4))


src/lib.js
import api from './api.js';

export default function preprocess(func) {
	let process = api();
	if(typeof func === "function") {
		func(process);
	} else if(typeof func === "string") {
		process.import(func);	
	} 
	return process;
}


// var test_html = {
//     div: 'hello world'
// };
//
// var test_css = {
//     div:{
//         margin: 0,
//         padding: 0,
//         position: 'absolute',
//         width: '224px',
//         height: '30px',
// 		'background-color': '#1a0a00',
// 		'padding-left': '5px',
//         '.ul_nav': {
// 			margin: 0,
// 			padding: 0,
// 			'list-style': 'none'
// 		},
//         '.ul_nav li': {
// 			margin: 0,
// 			padding: 0,
// 			display: "inline"
// 		},
//         '.link_group': {
//             margin: 0,
//             padding: 0,
//             display: 'inline-block',
// 			':hover':{
//                 '.link_nav': {
//                     color: '#ff6600'
//                 },
//                 '.nav_img': {
//                     filter: 'invert(92%) sepia(13%) saturate(682%) hue-rotate(311deg) brightness(107%) contrast(101%)'
//                 },
//                 '.nav_img_bg': {
//                     'background-color': '#1a0a00'
//                 }
// 			}
//         },
//         '.nav_img_bg': {
// 			margin: 0,
// 			padding: 0,
// 			width: '30px',
// 			height: '30px',
// 			'margin-right': '3px',
// 			'background-color': '#fff0e6',
// 			display: 'inline-block',
// 			'border-radius': '50%',
// 			'vertical-align': 'middle'
// 		},
//         '.nav_img': {
// 			margin: '6px',
// 			padding: 0,
// 			width: '18px',
// 			height: '18px',
//             filter: 'invert(6%) sepia(76%) saturate(641%) hue-rotate(345deg) brightness(95%) contrast(104%)'
// 		},
//         '.link_nav': {
// 			margin: 0,
// 			padding: 0,
// 			'text-decoration': 'none',
// 			color: '#fff0e6',
// 			'font-family': 'forced_squaremedium',
// 			'font-size': '28px',
// 			'margin-right': '3px',
// 			'vertical-align': 'middle',
//             outline: 'none',
//             ':active :focus': {
//                 outline: 'none'
//             }
// 		}
//     }
// };
//
// var test_component = {
//     html: test_html,
//     css: test_css
// };
//
// preprocess(function(api){
//     api.add(test_css).compile(function(err, result){
//         console.log(result);
//     });
//     api.morph('html').add(test_html).compile(function(err, result){
//         console.log(result);
//     });
//     api.add(test_component.css).compile(function(err, result){
//         console.log(result);
//     });
//     api.morph('html').add(test_component.html).compile(function(err, result){
//         console.log(result);
//     });
// });
